File: web-admin/api/device/[id]/preview.py
"""
Vercel serverless function for generating LED display previews.

This replaces the Node.js version that used child_process to spawn Python.
Instead, this native Python function directly imports and uses the preview generation code.
"""

from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import json
import os
import sys
from pathlib import Path
import tempfile
import uuid

# From web-admin/api/device/[id]/ go up 5 levels to reach rig root
sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent.parent))

from supabase import create_client
from src.config.supabase_config_loader import SupabaseConfigLoader, DeviceConfiguration
from src.preview.generator import PreviewGenerator


class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        try:
            parsed_url = urlparse(self.path)
            path_parts = parsed_url.path.split('/')
            device_id = path_parts[-2] if len(path_parts) >= 2 else None

            if not device_id:
                self._send_error(400, 'Device ID is required')
                return

            try:
                uuid.UUID(device_id)
            except (ValueError, AttributeError):
                self._send_error(400, f'Invalid device ID format: {device_id}')
                return

            auth_header = self.headers.get('Authorization', '')
            if not auth_header.startswith('Bearer '):
                self._send_error(401, 'Missing or invalid Authorization header')
                return

            access_token = auth_header[7:]

            supabase_url = os.getenv('NEXT_PUBLIC_SUPABASE_URL')
            supabase_anon_key = os.getenv('NEXT_PUBLIC_SUPABASE_ANON_KEY')

            if not supabase_url or not supabase_anon_key:
                self._send_error(500, 'Supabase configuration missing')
                return

            user_scoped = create_client(
                supabase_url,
                supabase_anon_key,
                options={
                    'headers': {'Authorization': f'Bearer {access_token}'},
                    'auth': {
                        'auto_refresh_token': False,
                        'persist_session': False
                    }
                }
            )

            auth_response = user_scoped.auth.get_user(access_token)
            if not auth_response or not auth_response.user:
                self._send_error(401, 'Unauthorized')
                return

            device_response = user_scoped.table('devices').select('id, user_id').eq('id', device_id).maybe_single().execute()

            if not device_response.data:
                self._send_error(403, 'Forbidden')
                return

            query_params = parse_qs(parsed_url.query)
            scene = query_params.get('scene', ['live'])[0]

            valid_scenes = ['idle', 'pregame', 'live', 'live_big', 'final']
            if scene not in valid_scenes:
                self._send_error(400, 'Invalid scene type')
                return

            service_role_key = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
            if not service_role_key:
                self._send_error(500, 'Service role key not configured')
                return

            service_client = create_client(supabase_url, service_role_key)
            loader = SupabaseConfigLoader(device_id, service_client)
            config = loader.load_full_config()

            if not config:
                self._send_error(404, 'Device configuration not found')
                return

            with tempfile.TemporaryDirectory() as output_dir:
                generator = PreviewGenerator(config, output_dir)

                if scene == "idle":
                    frame_path = generator.generate_idle_scene()
                elif scene == "pregame":
                    frame_path = generator.generate_pregame_scene(use_demo=True)
                elif scene == "live":
                    frame_path = generator.generate_live_scene(use_demo=True, big_logos=False)
                elif scene == "live_big":
                    frame_path = generator.generate_live_scene(use_demo=True, big_logos=True)
                elif scene == "final":
                    frame_path = generator.generate_final_scene(use_demo=True)
                else:
                    frame_path = generator.generate_live_scene(use_demo=True)

                with open(frame_path, 'rb') as f:
                    image_data = f.read()

                self.send_response(200)
                self.send_header('Content-Type', 'image/png')
                self.send_header('Cache-Control', 'no-store, must-revalidate')
                self.end_headers()
                self.wfile.write(image_data)

        except Exception as e:
            self._send_error(500, f'Failed to generate preview: {str(e)}')

    def _send_error(self, status_code, message):
        self.send_response(status_code)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        response = json.dumps({'error': message})
        self.wfile.write(response.encode('utf-8'))

File: web-admin/src/pages/api/device/[id]/preview.ts
import { NextApiRequest, NextApiResponse } from 'next'
import { createClient } from '@supabase/supabase-js'
import { exec } from 'child_process'
import { promisify } from 'util'
import path from 'path'
import fs from 'fs/promises'

const execAsync = promisify(exec)

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL!
const supabaseAnonKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
  const { id: deviceId } = req.query
  if (!deviceId || typeof deviceId !== 'string') {
    return res.status(400).json({ error: 'Device ID is required' })
  }

  const authHeader = req.headers.authorization || ''
  const tokenMatch = authHeader.match(/^Bearer\s+(.*)$/i)
  if (!tokenMatch) {
    return res.status(401).json({ error: 'Missing or invalid Authorization header' })
  }
  const accessToken = tokenMatch[1]

  const userScoped = createClient(supabaseUrl, supabaseAnonKey, {
    global: { headers: { Authorization: `Bearer ${accessToken}` } },
    auth: { autoRefreshToken: false, persistSession: false },
  })

  const { data: userData, error: authError } = await userScoped.auth.getUser(accessToken)
  if (authError || !userData?.user) {
    return res.status(401).json({ error: 'Unauthorized' })
  }

  const { data: deviceRow, error: deviceErr } = await userScoped
    .from('devices')
    .select('id, user_id')
    .eq('id', deviceId)
    .maybeSingle()

  if (deviceErr && deviceErr.code !== 'PGRST116') {
    return res.status(500).json({ error: deviceErr.message })
  }
  if (!deviceRow) {
    return res.status(403).json({ error: 'Forbidden' })
  }

  if (req.method === 'GET') {
    if (process.env.VERCEL || process.env.NETLIFY) {
      return res.status(503).json({
        error: 'Preview generation not available on serverless deployment',
        reason: 'serverless',
        details:
          'Preview generation requires Python runtime and file system access. ' +
          'For previews, deploy the web admin on a server with Python 3.8+ installed, ' +
          'or use a self-hosted environment. The preview feature will work on Railway, ' +
          'Render, or any VM-based hosting.',
      })
    }

    try {
      const scene = (req.query.scene as string) || 'live'
      const validScenes = ['idle', 'pregame', 'live', 'live_big', 'final']
      if (!validScenes.includes(scene)) {
        return res.status(400).json({ error: 'Invalid scene type' })
      }

      const projectRoot = path.resolve(process.cwd(), '..')
      const scriptPath = path.join(projectRoot, 'scripts', 'generate_preview.py')
      const outputDir = path.join(projectRoot, 'out', 'preview')

      const command = `cd ${projectRoot} && python3 ${scriptPath} --device-id ${deviceId} --scene ${scene} --output ${outputDir}`

      const { stdout, stderr } = await execAsync(command, {
        env: {
          ...process.env,
          PYTHONPATH: projectRoot,
        },
      })

      if (stderr) {
        console.error('Preview generation stderr:', stderr)
      }

      const result = JSON.parse(stdout)

      if (!result.success) {
        return res.status(500).json({ error: result.error || 'Preview generation failed' })
      }

      const framePath = result.path
      const imageBuffer = await fs.readFile(framePath)

      res.setHeader('Content-Type', 'image/png')
      res.setHeader('Cache-Control', 'no-store, must-revalidate')
      return res.send(imageBuffer)
    } catch (error: any) {
      console.error('Preview generation error:', error)
      return res.status(500).json({
        error: 'Failed to generate preview',
        details: error.message,
      })
    }
  }

  return res.status(405).json({ error: 'Method not allowed' })
}
